libyul/YulName.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/YulName.h>
+
+#include <libyul/Exceptions.h>
+
+#include <fmt/compile.h>
+
+namespace solidity::yul
+{
+
+YulNameRepository::YulNameRepository()
+{
+	auto const emptyName = defineName("");
+	yulAssert(emptyName == YulNameRepository::emptyName());
+}
+
+std::string_view YulNameRepository::requiredLabelOf(YulName const _name) const
+{
+	auto const label = labelOf(_name);
+	yulAssert(label.has_value(), "YulName currently has no defined label in the YulNameRepository.");
+	return label.value();
+}
+
+std::optional<YulNameRepository::YulName> YulNameRepository::nameOfLabel(std::string_view const _label) const
+{
+	auto const it = std::find(m_definedLabels.begin(), m_definedLabels.end(), _label);
+	if (it != m_definedLabels.end())
+	{
+		YulName labelName{static_cast<YulName>(std::distance(m_definedLabels.begin(), it))};
+		// mostly it'll be iota
+		if (!isDerivedName(labelName) && std::get<0>(m_names[static_cast<size_t>(labelName)]) == labelName)
+			return std::get<0>(m_names[static_cast<size_t>(labelName)]);
+		// if not iota, we have to search
+		auto itName = std::find(m_names.rbegin(), m_names.rend(), std::make_tuple(labelName, YulNameState::DEFINED));
+		if (itName != m_names.rend())
```

Comment:
Can a label exist without a name pointing at it?
```suggestion
		yulAssert(itName != m_names.rend());
```